src/fomo.jl

```diff
@@ -0,0 +1,399 @@
+export fomo, FomoSolver, FoSolver, R2, tr_step, r2_step
+
+abstract type AbstractFirstOrderSolver <: AbstractOptimizationSolver end
+
+abstract type AbstractFomoMethod end
+struct tr_step   <: AbstractFomoMethod end
+struct r2_step   <: AbstractFomoMethod end
+
+"""
+    fomo(nlp; kwargs...)
+    R2(nlp; kwargs...)
+
+A First-Order with MOmentum (FOMO) model-based method for unconstrained optimization. Supports quadratic regularization and trust region methods.
+
+For advanced usage, first define a `FomoSolver` to preallocate the memory used in the algorithm, and then call `solve!`:
+
+    solver = FomoSolver(nlp)
+    solve!(solver, nlp; kwargs...)
+
+**Quadratic Regularization (R2)**: if the user do not want to use momentum (`β` = 0), it is recommended to use the memory-optimized `R2` method.
+For advanced usage:
+
+    solver = FoSolver(nlp)
+    solve!(solver, nlp; kwargs...)
+Extra keyword arguments `σmin` is accepted (`αmax` will be set to `1/σmin`).
+
+# Arguments
+- `nlp::AbstractNLPModel{T, V}` is the model to solve, see `NLPModels.jl`.
+
+# Keyword arguments 
+- `x::V = nlp.meta.x0`: the initial guess.
+- `atol::T = √eps(T)`: absolute tolerance.
+- `rtol::T = √eps(T)`: relative tolerance: algorithm stops when ‖∇f(xᵏ)‖ ≤ atol + rtol * ‖∇f(x⁰)‖.
+- `η1 = eps(T)^(1/4)`, `η2 = T(0.95)`: step acceptance parameters.
+- `γ1 = T(1/2)`, `γ2 = T(2)`: regularization update parameters.
+- `γ3 = T(1/2)` : momentum factor βmax update parameter in case of unsuccessful iteration.
+- `αmax = 1/eps(T)`: maximum step parameter for fomo algorithm.
+- `max_eval::Int = -1`: maximum number of evaluation of the objective function.
+- `max_time::Float64 = 30.0`: maximum time limit in seconds.
+- `max_iter::Int = typemax(Int)`: maximum number of iterations.
+- `β = T(0.9) ∈ [0,1)` : target decay rate for the momentum.
+- `θ1 = T(0.1)` : momentum contribution parameter for convergence condition #1. (1-βmax) * ∇f(xk) + βmax * dot(m,∇f(xk)) ≥ θ1 * ‖∇f(xk)‖², with m memory of past gradient and βmax ∈ [0,β].
+- `θ2::T = T(eps(T)^(1/3))` : momentum contribution parameter for convergence condition #2. ‖∇f(xk)‖ ≥ θ2 * ‖(1-βmax) * ∇f(xk) + βmax * m‖, with m memory of past gradient and βmax ∈ [0,β]. 
+- `verbose::Int = 0`: if > 0, display iteration details every `verbose` iteration.
+- `step_backend = r2_step()`: step computation mode. Options are `r2_step()` for quadratic regulation step and `tr_step()` for first-order trust-region.
+
+# Output
+The value returned is a `GenericExecutionStats`, see `SolverCore.jl`.
+
+# Callback
+The callback is called at each iteration.
+The expected signature of the callback is `callback(nlp, solver, stats)`, and its output is ignored.
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
```

Comment:
Ce n’est pas tout à fait vrai. L’exécution s’arrête si `status != :unknown`.

Reply:
This is a copy paste from former R2.jl. Comment has been updated.